ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1721,6 +1759,484 @@ void shutdown() throws InterruptedException {
+      logger.error("Remote node: {} Thread name: {} RemoteReplicaGroup {} {}", remoteDataNode, threadName, id, message,
+          e);
+      exception = e;
+      state = ReplicaGroupReplicationState.DONE;
+    }
+
+    /**
+     * @return The exception.
+     */
+    public Exception getException() {
+      return exception;
+    }
+  }
+
+  /**
+   * Do replication for replicas grouped by {@link DataNodeId} with nonblocking network client.
+   */
+  public void replicateWithNetworkClient() {
+    exchangeMetadataResponsesInEachCycle = new HashMap<>();
+    long oneRoundStartTimeMs = time.milliseconds();
+    logger.trace("Thread name: {} Start RemoteReplicaGroup replication", threadName);
+    List<RemoteReplicaGroup> remoteReplicaGroups = new ArrayList<>();
+    int remoteReplicaGroupId = 0;
+
+    // Before each cycle of replication, we clean up the cache in key converter.
+    storeKeyConverter.dropCache();
+    Map<DataNodeId, List<RemoteReplicaInfo>> dataNodeToRemoteReplicaInfo = getRemoteReplicaInfos();
+    for (Map.Entry<DataNodeId, List<RemoteReplicaInfo>> entry : dataNodeToRemoteReplicaInfo.entrySet()) {
+      DataNodeId remoteNode = entry.getKey();
+      List<RemoteReplicaInfo> replicasToReplicatePerNode = entry.getValue();
+      List<RemoteReplicaInfo> activeReplicasPerNode = new ArrayList<>();
+      List<RemoteReplicaInfo> standbyReplicasWithNoProgress = new ArrayList<>();
+      filterRemoteReplicasToReplicate(replicasToReplicatePerNode, activeReplicasPerNode, standbyReplicasWithNoProgress);
+
+      if (activeReplicasPerNode.size() > 0) {
+        List<List<RemoteReplicaInfo>> activeReplicaSubLists =
+            maxReplicaCountPerRequest > 0 ? Utils.partitionList(activeReplicasPerNode, maxReplicaCountPerRequest)
+                : Collections.singletonList(activeReplicasPerNode);
+        for (List<RemoteReplicaInfo> replicaSubList : activeReplicaSubLists) {
+          RemoteReplicaGroup group = new RemoteReplicaGroup(replicaSubList, remoteNode, false, remoteReplicaGroupId++);
+          remoteReplicaGroups.add(group);
+        }
+      }
+      if (standbyReplicasWithNoProgress.size() > 0) {
+        List<RemoteReplicaInfo> standbyReplicasTimedOutOnNoProgress =
+            getRemoteStandbyReplicasTimedOutOnNoProgress(standbyReplicasWithNoProgress);
+        RemoteReplicaGroup group =
+            new RemoteReplicaGroup(standbyReplicasTimedOutOnNoProgress, remoteNode, true, remoteReplicaGroupId++);
+        remoteReplicaGroups.add(group);
+      }
+    }
+    // A map from correlation id to RemoteReplicaGroup. This is used to find the group when response comes back.
+    Map<Integer, RemoteReplicaGroup> correlationIdToReplicaGroup = new HashMap<>();
+    // A map from correlation id to RequestInfo. This is used to find timed out RequestInfos.
+    Map<Integer, RequestInfo> correlationIdToRequestInfo = new LinkedHashMap<>();
+    while (!remoteReplicaGroups.stream().allMatch(RemoteReplicaGroup::isDone)) {
```

Comment:
Just a question. Since we are waiting for replication to complete across all hosts for this cycle, if one of the remote host is slow or has connection problems, would it slow replication for all hosts. Just thinking if this similar to what we see during Threads.join().

Reply:
That's right, and that's the design for now.  We are having two implementation for replication, one with blocking channel and the other with nonblocking network client. I want to make sure that we can keep these two implementations during testing for nonblocking network client. So I have to do one cycle replication at a time.

After we remove the blocking channel and only use nonblocking network client, we can then replicate same RemoteReplicaGroup at different cycle without joining.